DEV-17017 Reorganize and simplify dependencies

Maldron opened the pull request:
First pass at this, kinda messy but working and way nicer import API.

Move standard C FFI functionality to `ffi_core`. Strip everything else
out of `ffi_common`; repurpose it as a top-level crate that reexports
`ffi_core`, `ffi_consumer`, and `ffi_derive`.

Update proc macros to use reexports from `ffi_common`. Picked up a fun
trick from serde for tests -- `ffi_derive` can have a dev-dependency on
`ffi_common`, which will let the tests live in `ffi_derive` while still
using the `ffi_common` import paths.

With these changes, consumer crates only have to depend on `ffi_common`
(and they still don't have access to any of the token parsing madness in
`ffi_internals`).

Maldron committed:
DEV-17017 Reorganize and simplify dependencies

First pass at this, kinda messy but working and way nicer import API.

Move standard C FFI functionality to `ffi_core`. Strip everything else
out of `ffi_common`; repurpose it as a top-level crate that reexports
`ffi_core`, `ffi_consumer`, and `ffi_derive`.

Update proc macros to use reexports from `ffi_common`. Picked up a fun
trick from serde for tests -- `ffi_derive` can have a dev-dependency on
`ffi_common`, which will let the tests live in `ffi_derive` while still
using the `ffi_common` import paths.

With these changes, consumer crates only have to depend on `ffi_common`
(and they still don't have access to any of the token parsing madness in
`ffi_internals`).

Maldron committed:
More dependency reexports from ffi_internals.

Also enable `extra-traits` for `syn`. I'm not sure how we were getting
away with `Debug` and `thiserror` impls for `syn` types before, since
those are gated behind this feature, but `wise_units` just started
noticing the issue when building with this branch.

turboladen commented on ffi_common/build.rs:
```diff
@@ -1,17 +1,2 @@
-            false
-        });
-}
+fn main() { }
```
I forget--do you still need this for anything?

Maldron replied:
Nope! Had to take another look, but only `ffi_derive` and `ffi_internals` need a `build.rs`, so this can go away. The output directory is passed in everywhere else.

Also gonna drop the `cbindgen` build dependency -- the docs already have notes on how to run that from the consumer crate, where it can create a header for the whole thing. It just stuck around from earlier versions.

Maldron committed:
Nuke unnecessary `build.rs` files.

Remove unnecessary cbindgen build dependency and `ffi_internals` crate-type specification.

Maldron committed:
Update GH action to stop running tests twice on PRs.
